virtcontainers/container.go

```diff
@@ -1076,18 +1111,30 @@ func (c *Container) removeDrive() (err error) {
 }
 
 func (c *Container) attachDevices() error {
-	for _, device := range c.devices {
-		if err := device.Attach(c.sandbox); err != nil {
+	for _, dev := range c.devices {
+		if err := c.sandbox.devManager.AttachDevice(dev.ID, c.sandbox); err != nil {
+			if err == manager.ErrDeviceAttached {
+				// skip if device is already attached before
+				continue
+			}
 			return err
 		}
 	}
 
+	if err := c.sandbox.storeSandboxDevices(); err != nil {
+		//TODO: roll back?
+		return err
+	}
 	return nil
 }
 
 func (c *Container) detachDevices() error {
-	for _, device := range c.devices {
-		if err := device.Detach(c.sandbox); err != nil {
+	for _, dev := range c.devices {
```

Comment:
This function is identical to `attachDevices()` except for the different API call. It might be worth creating a `toggleAttachDevices(attach bool)` function that both `attachDevices()` and `detachDevices()` call to avoid the duplication and ensure they remain consistent.

Reply:
Worth a try, I'll modify as per your suggestion.
Thank @jodh-intel for helping review!